File: src/move_linear.cpp
#include <string.h>
#include <ctime>
#include <iostream>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"

using std::placeholders::_1;
using namespace std;
using namespace std::chrono_literals;

class MoveLinear : public rclcpp::Node
{
    private:
        rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr speed_pub_;
        geometry_msgs::msg::Twist mov;
        time_t start, finish;
        
    public:
        MoveLinear() : Node("move_linear"){
            speed_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
                "turtle1/cmd_vel", 10);
        }

        void move_distance(float speed, float distance, int direction){
            float expected_time = distance/speed;
            time(&start);

            if(direction == 0){
                mov.linear.x = -1 * speed;
            } else {
                mov.linear.x = speed;
            }

            while(difftime(time(&finish), start) < expected_time){
                // cout << difftime(time(&finish), start) << " " << expected_time << endl;
                speed_pub_->publish(mov);
            }
        }
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);

    MoveLinear MvLinaer;

    float speed;
    float distance;
    float direction;

    cout << "Enter speed: ";
    cin >> speed;

    cout << "Enter Distance: ";
    cin >> distance;

    cout << "Enter direction (1-Forward, 0-Backward): ";
    cin >> direction;


    MvLinaer.move_distance(speed, distance, direction); 

    rclcpp::shutdown();
    return 0;
}

File: scripts/move_linear.py
#!/usr/bin/env python3

import rclpy
import time

from math import pi
from rclpy.node import Node
from geometry_msgs.msg import Twist


class MoveLinear(Node):

    def __init__(self):
        super().__init__("move_linear") # type: ignore
        self.speed_pub_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.prev_time = 0

    def move_distance(self, speed, distance):
        mov = Twist()

        mov.linear.x = float(speed)

        expected_time = distance / speed
        print(expected_time)

        start_time = int(time.time())
        
        while(self.time_gone(start_time) < expected_time):
            self.speed_pub_.publish(mov)
    
    def backward(self):
        rot = Twist()

        direction = pi
        angular_velocity = 10.0

        rot.angular.z = direction   
        rot.angular.x = angular_velocity
        
        self.speed_pub_.publish(rot)

        time.sleep(0.1)
        
    def time_gone(self, start):
        time_now = int(time.time())
        time_gone = time_now - start

        return time_gone
        
def main(args = None):
    rclpy.init(args=args)
    
    MvLinaer = MoveLinear()

    speed = int(input("Enter speed: "))
    distance = int(input("Enter Distance speed: "))
    direction = int(input("Enter direction (1-Forward, 0-Backward): "))
    
    if (direction == 0):
        MvLinaer.backward()

    MvLinaer.move_distance(speed, distance)

    rclpy.shutdown()

if __name__ == '__main__':
    main()